pricegraph/src/orderbook.rs

```diff
@@ -183,21 +183,90 @@ impl Orderbook {
         overlap
     }
 
+    /// Fills transitive orders along a token pair, optionally specifying a
+    /// maximum price spread for the orders.
+    ///
+    /// Returns a vector containing all the transitive orders that were filled.
+    ///
+    /// Note that the spread is a decimal fraction that defines the maximum
+    /// transitive order price with the equation:
+    /// `first_transitive_price + first_transitive_price * spread`. This means
+    /// that given a spread of 0.5 (or 50%), and if the cheapest transitive
+    /// order has a price of 1.2, then the maximum price will be `1.8`.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if the spread is invalid, that is not in the range
+    /// `(0.0, +∞)`.
+    pub fn fill_transitive_orders(
+        &mut self,
+        pair: TokenPair,
+        spread: Option<f64>,
+    ) -> Vec<TransitiveOrder> {
+        if let Some(spread) = spread {
+            assert!(spread > 0.0, "invalid spread");
+        }
+
+        if !self.is_token_pair_valid(pair) {
+            return Vec::new();
+        }
+        self.update_projection_graph();
+
+        let (sell, buy) = (node_index(pair.sell), node_index(pair.buy));
+
+        let mut orders = Vec::new();
+        let mut spread_limit_price = None;
+
+        // NOTE: A subtle difference here is that we are searching for paths in
+        // the same direction as the token pair (as opposed to the opposite
+        // direction when doing price estimates).
+        let mut predecessors = self.reduced_shortest_paths(buy);
+
+        while let Some(path) = path::find_path(&predecessors, buy, sell) {
```

Comment:
nit: if you are thinking about refactoring the code, we might be able to wrap the `while let` + `fill_path_with_capacity` + `predecessors = bellman_ford::search` in a `forEachPath` method.

Reply:
Yes! Yes! Yes! This one of the current refactors I am working on now. There are two of these loops:
```rust
let mut predecessors = reduced_shortest_path(...);
while let Some(path) = path::find_path(predecessors, ...) {
  let (capacity, price) = fill_path(path);
  if !condition(capacity, price) {
    break;
  }
  predecessors = bellman_ford::search(...);
}
```

And a super similar:
```
let mut predecessors = reduced_shortest_path(...);
while let Some(path) = path::find_path(predecessors, ...) {
  let (capacity, price) = find_path_capacity_and_price(path);
  if !condition(capacity, price) {
    break;
  }
  fill_path_with_capacity(path, capacity)
  predecessors = bellman_ford::search(...);
}
```

Basically in one loop, you check the condition before filling, and in the other you check the condition after filling. All other things are the same (including updating the projection, checking for token validity, etc.).

This is the next PR. And then I want to move some code around since this file is over 1000 lines already :see_no_evil: